webapp/channels/src/packages/mattermost-redux/src/reducers/entities/users.ts

```diff
@@ -173,65 +173,81 @@ function myAudits(state = [], action: GenericAction) {
     }
 }
 
+function receiveUserProfile(state: IDMappedObjects<UserProfile>, received: UserProfile) {
+    const existing = state[received.id];
+
+    if (!existing) {
+        // No existing data to merge with
+        return {
+            ...state,
+            [received.id]: received,
+        };
+    }
+
+    const merged = {
+        ...existing,
+        ...received,
+    };
+
+    // MM-53377:
+    // For non-admin users, certain API responses don't return details for the current user that would be sanitized
+    // out for others. This currently includes:
+    // - email (if PrivacySettings.ShowEmailAddress is false)
+    // - first_name/last_name (if PrivacySettings.ShowFullName is false)
+    // - last_password_update
+    // - auth_service
+    // - notify_props
+    //
+    // Because email, first_name, last_name, and auth_service can all be empty strings regularly, we can't just
+    // merge the received user and the existing one together like we normally would. Instead, we can use the
+    // existence of existing.notify_props or existing.last_password_update to determine which object has that extra
+    // data so that it can take precedence. Those fields are:
+    // 1. Never empty or zero by Go standards
+    // 2. Only ever sent to the current user, not even to admins, so we know that the object contains privileged data
+    //
+    // Note that admins may have the email/name/auth_service of other users loaded as well. This does not prevent that
+    // data from being replaced when merging sanitized user objects. There doesn't seem to be a way for us to detect
+    // whether the object is sanitized for admins.
```

Comment:
No, unfortunately not. Admins are usually sent unsanitized objects, but there's at least some channel-wide or team-wide websocket traffic that is always sent as sanitized.

The admin case is an existing issue that wasn't introduced by the reverted changes, and often admin-related code reloads the User objects manually (like in the System Console Users list), so I decided to leave it as-is. A more ideal fix for this would be to either fix all of our sanitization code to be more consistent (hard) or to add something like a field on the object for the frontend code to more clearly understand the sanitization level (less hard but still a fair bit of work)